AdminServer/appscale/admin/service_manager.py

```diff
@@ -0,0 +1,410 @@
+    # This is for compatibility with Hermes, which expects a monit name.
+    self.monit_name = None
+    self.port = port
+    self.process = None
+    self.state = ServerStates.NEW
+    self.type = service_type
+
+  @gen.coroutine
+  def start(self):
+    raise NotImplementedError()
+
+  @gen.coroutine
+  def stop(self):
+    raise NotImplementedError()
+
+  def __repr__(self):
+    """ Represents the service details.
+
+    Returns:
+      A string representing the service.
+    """
+    return '<Service: {}:{}, {}>'.format(self.type, self.port, self.state)
+
+
+class DatastoreServer(Server):
+  """ Keeps track of the status and location of a datastore server. """
+
+  # The datastore backend.
+  DATASTORE_TYPE = 'cassandra'
+
+  # The number of seconds to wait for the server to start.
+  START_TIMEOUT = 30
+
+  # The number of seconds to wait for a status check.
+  STATUS_TIMEOUT = 10
+
+  # The number of seconds to wait for the server to stop.
+  STOP_TIMEOUT = 5
+
+  def __init__(self, port, http_client, verbose):
+    """ Creates a new DatastoreServer.
+
+    Args:
+      port: An integer specifying the port to use.
+      http_client: An AsyncHTTPClient
+      verbose: A boolean that sets logging level to debug.
+    """
+    super(DatastoreServer, self).__init__(ServiceTypes.DATASTORE, port)
+    self.monit_name = 'datastore_server-{}'.format(port)
+    self._http_client = http_client
+    self._stdout = None
+    self._verbose = verbose
+
+  @gen.coroutine
+  def ensure_running(self):
+    """ Checks to make sure the server is still running. """
+    yield self._wait_for_service(timeout=self.STATUS_TIMEOUT)
+
+  @classmethod
+  def from_pid(cls, pid, http_client):
+    """ Creates a new DatastoreServer from an existing process.
+
+    Args:
+      pid: An integers specifying a process ID.
+      http_client: An AsyncHTTPClient.
+    """
+    process = psutil.Process(pid)
+    args = process.cmdline()
+    port = int(args[args.index('--port') + 1])
+    verbose = '--verbose' in args
+    server = cls(port, http_client, verbose)
+    server.process = process
+    server.state = ServerStates.RUNNING
+    return server
+
+  @gen.coroutine
+  def start(self):
+    """ Starts a new datastore server. """
+    if self.state in (ServerStates.STARTING, ServerStates.RUNNING):
+      raise gen.Return()
+
+    self.state = ServerStates.STARTING
+    start_cmd = ['appscale-datastore',
+                 '--type', self.DATASTORE_TYPE,
+                 '--port', str(self.port)]
+    if self._verbose:
+      start_cmd.append('--verbose')
+
+    log_file = os.path.join(LOG_DIR,
+                            'datastore_server-{}.log'.format(self.port))
+    self._stdout = open(log_file, 'a')
+    self.process = psutil.Popen(
+      ['cgexec', '-g', ':'.join(DATASTORE_CGROUP)] + start_cmd,
+      stdout=self._stdout, stderr=subprocess.STDOUT)
+
+    # Wait for server to bind to port before making HTTP requests.
+    yield gen.sleep(1)
```

Comment:
The problem I ran into during testing is that if the http client was in the process of trying a request while the server started, the server could not bind to the port. The ServiceManager would eventually get a server started, but it was kind of annoying to have failed starts. Do you think it would be better to be more persistent with the server to try to bind to port on startup?

Reply:
I didn't know that requesting port can affect server ability to bind port.
Adding retries on binding port process with reasonable timeout sounds good to me.
Though it can go as a separate issue out of this PR.